lib/cs_guide/postcode_lat_long.ex

```diff
@@ -0,0 +1,132 @@
+defmodule CsGuide.PostcodeLatLong do
+  alias NimbleCSV.RFC4180, as: CSV
+  alias CsGuide.Resources.Venue
+  import Ecto.Query
+  use GenServer
+
+  @zip_file_name "postcodes.zip"
+  @zip_file_name_charlist String.to_charlist(@zip_file_name)
+  @postcode_file_charlist 'ukpostcodes.csv'
+
+  def start_link do
+    GenServer.start_link(__MODULE__, %{}, name: PostcodeCache)
+  end
+
+  def init(initial_state) do
+    postcode_cache = :ets.new(:postcode_cache, [:set, :protected, :named_table])
+    store_postcodes_in_ets(postcode_cache)
+    {:ok, initial_state}
+  end
+
+  defp store_postcodes_in_ets(ets) do
+    if File.exists?("ukpostcodes.csv") do
+      IO.inspect("Postcode file exists, storing with ets")
+      get_postcodes_from_csv("ukpostcodes.csv", ets)
+    else
+      IO.inspect("Postcode file does not exist. Creating file...")
+      case create_csv_file() do
+        :ok ->
+          IO.inspect("Removed zip file. Storing postcodes with ets")
+          get_postcodes_from_csv("ukpostcodes.csv", ets)
+
+        error ->
+          error
+      end
+    end
+  end
+
+  def nearest_venues(lat, long, distance \\ 5_000)
+  def nearest_venues(lat, long, distance) when distance < 30_000 do
+    case venues_within_distance(distance, lat, long) do
+      [] -> nearest_venues(lat, long, distance + 5_000)
+      venues -> venues
+    end
+  end
+  def nearest_venues(_, _, _), do: []
+
+  defp venues_within_distance(distance, lat, long) do
+    Venue
+    |> where([venue], venue.deleted == false)
+    |> where(
+      [venue],
+      fragment(
+        "? @> ?",
+        fragment(
+          "earth_box(?, ?)",
+          fragment("ll_to_earth(?, ?)", ^lat, ^long),
+          ^distance
+        ),
+        fragment("ll_to_earth(?,?)", venue.lat, venue.long)
+      )
+    )
+    |> select([venue], %{
+      venue
+      | distance:
+          fragment(
+            "? as distance",
+            fragment(
+              "? <@> ?",
+              fragment("point(?, ?)", ^long, ^lat),
+              fragment("point(?, ?)", venue.long, venue.lat)
+            )
+          )
+    })
+    |> order_by(desc: :inserted_at)
+    |> distinct([v], fragment("distance"))
```

Comment:
I think distinct on `entry_id` would be better than distance. If two venues happened to be on the same road, they would also have the same postcode, and therefore the same distance, so one of them would be filtered out.

Reply:
That's a good point. I wasn't sure quite how the distance filtered addressed so and as it was using lat long values to calculate distance I thought that 2 venues on the same road (even with the same postcode) would have different distances. 

I think you're right though and it should be done on `:entry_id`. I will look into changing that now but I think it could cause issues with the order_by functionality.

Thanks @Danwhy 👍

Reply:
`distinct([v], :entry_id)` is causing the `order_by` to stop working

People seem to have run into this order_by and distinct problem before. https://weblogs.sqlteam.com/jeffs/2007/12/13/select-distinct-order-by-error/

Reading the above to see if I can try and better understand what is going wrong here

Reply:
https://github.com/club-soda/club-soda-guide/pull/291/commits/5b178a3df749ccd209486441bde76764ca7b0814 @Danwhy updated now 👍